tests/integration/test_migrate.py

```diff
@@ -135,9 +136,88 @@ def test_happy_migrate(cluster_factory):
     assert cluster2.execute("get", "key1") is None
 
+
+def test_sad_path_migrate(cluster_factory):
+    cluster1 = cluster_factory().create(1, raft_args={
+        'sharding': 'yes',
+        'external-sharding': 'yes'})
+    cluster2 = cluster_factory().create(1, raft_args={
+        'sharding': 'yes',
+        'external-sharding': 'yes'})
+
+    cluster1_dbid = cluster1.leader_node().info()["raft_dbid"]
+    cluster2_dbid = cluster2.leader_node().info()["raft_dbid"]
+
+    assert cluster1.execute('set', 'key1', 'value1')
+    assert cluster1.execute('get', 'key1') == b'value1'
+    assert cluster1.execute('set', 'key2', 'value2')
+    assert cluster1.execute('get', 'key2') == b'value2'
+    assert cluster1.execute('set', 'key3', 'value3')
+    assert cluster1.execute('get', 'key3') == b'value3'
+
+    assert cluster1.execute(
+        'RAFT.SHARDGROUP', 'REPLACE',
+        '2',
+        cluster2_dbid,
+        '1', '1',
+        '0', '16383', '2', '123',
+        '%s00000001' % cluster2_dbid, 'localhost:%s' % cluster2.node(1).port,
+        cluster1_dbid,
+        '1', '1',
+        '0', '16383', '3', '123',
+        '%s00000001' % cluster1_dbid, 'localhost:%s' % cluster1.node(1).port,
+        ) == b'OK'
+
+    assert cluster2.execute(
+        'RAFT.SHARDGROUP', 'REPLACE',
+        '2',
+        cluster2_dbid,
+        '1', '1',
+        '0', '16383', '2', '123',
+        '%s00000001' % cluster2_dbid, 'localhost:%s' % cluster2.node(1).port,
+        cluster1_dbid,
+        '1', '1',
+        '0', '16383', '3', '123',
+        '%s00000001' % cluster1_dbid, 'localhost:%s' % cluster1.node(1).port,
+        ) == b'OK'
+
+    def validate_failed_migration(key_name, value, slot, err_string):
+        # first pass, should error out
+        with raises(ResponseError, match=err_string):
+            cluster1.execute("migrate", "", "", "", "", "", "keys", key_name)
+
+        # validate state
+        with raises(ResponseError, match="TRYAGAIN"):
+            cluster1.execute("get", key_name)
+        with raises(ResponseError, match=f"MOVED {slot} localhost"):
+            cluster2.leader_node().client.get(key_name)
+
+        # remove injected error, should pass
+        cluster1.execute("raft.debug", "migration_debug", 0)
+        assert cluster1.execute("migrate", "", "", "", "", "", "keys", key_name) == b'OK'
+
+        # validate state
+        with raises(ResponseError, match=f"ASK {slot} localhost"):
+            cluster1.execute("get", key_name)
+        with raises(ResponseError, match=f"MOVED {slot} localhost"):
+            cluster2.leader_node().client.get(key_name)
+
+        conn = cluster2.leader_node().client.connection_pool.get_connection('deferred')
+        conn.send_command('ASKING')
+        assert conn.read_response() == b'OK'
+        conn.send_command('get', key_name)
+        assert conn.read_response() == value
+
+    cluster1.execute("raft.debug", "migration_debug", 1)
```

Comment:
```suggestion
    cluster1.execute("raft.debug", "migration_debug", 'fail_connect')
```

```c
        int val;
        
        size_t len;
        const char *str = RedisModule_StringPtrLen(argv[2], &len);
        if (!strncasecmp(str, "fail_connect", len)) {
            val = DEBUG_MIGRATION_EMULATE_CONNECT_FAILED;
        } else if (!strncasecmp(str, "fail_import", len)) {
            val = DEBUG_MIGRATION_EMULATE_IMPORT_FAILED;
        } else if (!strncasecmp(str, "fail_unlock", len)) {
            val = DEBUG_MIGRATION_EMULATE_UNLOCK_FAILED;
        } else {
            RedisModule_ReplyWithError(ctx, "ERR invalid migration debug value");
            return REDISMODULE_OK;
        }
```

You can accept strings if you want, so it'd be easier to understand inside the test file. Just an idea.

Reply:
I didn't try the code btw, fyi